### ISSUE TITLE
Unable to build, fails while building melange-compiler-libs

### ISSUE THREAD

**Issue by Ilazki** [melange-re/melange#202]
I've been banging my head against this with no luck for days on and off.  Whenever I try to build melange on my desktop (Debian stable, amd64), either directly (by cloning this repo and running `esy`) or indirectly (cloning melange-basic-template and running `esy`), it always fails when attempting to build melange-compiler-libs because of a problem with dot-merlin-reader:

```
$ esy
info esy 0.6.10 (using esy.json)
info building @opam/reason@opam:3.7.0@494dd52d
info building @opam/melange-compiler-libs@github:melange-re/melange-compiler-libs:melange-compiler-libs.opam#c787d2f98a@d41d8cd9
info building @opam/dot-merlin-reader@opam:4.1@84436e1c
error: build failed with exit code: 1
  build log:
    # esy-build-package: building: @opam/dot-merlin-reader@opam:4.1
    # esy-build-package: pwd: /home/melange/.esy/3/b/opam__s__dot_merlin_reader-opam__c__4.1-cd660a35
    # esy-build-package: running: 'dune' 'build' '-p' 'dot-merlin-reader' '-j' '4'
        ocamlopt src/dot-merlin/dot_merlin_reader.exe (exit 2)
    (cd _build/default && /home/melange/.esy/3_________________________________________________________________/i/ocaml-4.12.0-bfdfbfff/bin/ocamlopt.opt -w -40 -g -o src/dot-merlin/dot_merlin_reader.exe /home/melange/.esy/3_________________________________________________________________/i/opam__s__ocamlfind-opam__c__1.9.1-6131b28f/lib/findlib/findlib.cmxa /home/melange/.esy/3_________________________________________________________________/i/opam__s__easy_format-opam__c__1.3.2-df2814bf/lib/easy-format/easy_format.cmxa /home/melange/.esy/3_________________________________________________________________/i/opam__s__biniou-opam__c__1.2.1-4ec85bf9/lib/biniou/biniou.cmxa /home/melange/.esy/3_________________________________________________________________/i/opam__s__yojson-opam__c__1.7.0-1dce999c/lib/yojson/yojson.cmxa /home/melange/.esy/3_________________________________________________________________/i/opam__s__result-opam__c__1.5-5675190b/lib/result/result.cmxa src/utils/merlin_utils.cmxa -I src/utils /home/melange/.esy/3_________________________________________________________________/i/opam__s__csexp-opam__c__1.5.1-c6a35626/lib/csexp/csexp.cmxa src/dot-merlin/dot-protocol/merlin_dot_protocol.cmxa src/dot-merlin/.dot_merlin_reader.eobjs/native/dune__exe__Dot_merlin_reader.cmx)
    File "_none_", line 1:
    Error: No implementations provided for the following modules:
             Str referenced from src/utils/merlin_utils.cmxa(Merlin_utils__Std)
             Unix referenced from src/utils/merlin_utils.cmxa(Merlin_utils__File_id),
               src/utils/merlin_utils.cmxa(Merlin_utils__File_cache),
               src/utils/merlin_utils.cmxa(Merlin_utils__Misc),
               src/dot-merlin/.dot_merlin_reader.eobjs/native/dune__exe__Dot_merlin_reader.cmx
    error: command failed: 'dune' 'build' '-p' 'dot-merlin-reader' '-j' '4' (exited with 1)
    esy-build-package: exiting with errors above...

  building @opam/dot-merlin-reader@opam:4.1
esy: exiting due to errors above
```

Even with a brand new user account, no existing .opam or .esy or anything else installed, it still fails.  However, it builds fine on a VPS I have that's also running amd64 Debian stable and is configured similarly to my desktop, so I'm baffled.

**Comment by EduardoRFS** [comment]
The problem seems related to `dot-merlin-reader`, as you can see in

```
    esy-build-package: exiting with errors above...

  building @opam/dot-merlin-reader@opam:4.1
  ```

**Comment by Ilazki** [comment]
Yes, I mentioned that above:  " it always fails when attempting to build melange-compiler-libs because of a problem with dot-merlin-reader".

However it's **only** failing in this one specific part of melange compilation.  It appears to get built more than once as a dependency of other packages just fine, but the one after the `@opam/reason` and `@opam/melange-compiler-libs` messages **always** fails, but only as part of the `esy` build process.  I can install merlin (which uses dot-merlin-reader as a dependency) manually without issue, and I can even go into the listed on that's failing (`/home/melange/.esy/3/b/opam__s__dot_merlin_reader-opam__c__4.1-cd660a35`) and build it manually.  But it fails when being built via `esy` which makes melange unbuildable and unusable.

So yes, it's dot-merlin-reader failing, but the problem is still with the melange build process somewhere but I don't know why, which is why I reported it here.

**Comment by Ilazki** [comment]
Follow up: I came back to trying to figure this out and managed to solve the problem, though I still don't quite know why it was a problem at all.  After eliminating user and shell differences (it failed on my system on a new user with bash but built fine in a VM with a near-identical user setup using my normal shell/configs/etc.) I figured it had to be a system-level problem and started checking for differences.

The problem was that part of the melange build process was erroneously trying to use the system-wide ocaml from my distro's package manager (which was installed as a dependency for another piece of software) and then failing for some reason.  Still no clue why or what was going on, since everything worked fine *except* the specific combination of building Melange using esy on the system, but once I removed that everything started working as expected.

Since it's a very edge-case problem and I found a workaround, I'm just going to close the bug because there's not much reason to dig into what exactly went wrong here; it would just be a waste of time and resources at this point.  But, in case anybody in the future encounters a similar problem and finds this bug report, I'm leaving this comment to describe the basic problem and what fixed it for me.
